Added action buttons to the datagrid

ryan-bradford opened the pull request:
# Description

Adds action buttons to the datagrid. There are two types of buttons - 

1. Global Buttons, which always appear at the top of the datagrid and whose visibility does not depend on the row currently selected.
2. Contextual buttons, which can appear either at the top of the grid or next to each row, and whose visibility does depend on the row(s) currently selected.

There are also various options for visibility:

1. Hide: When the button is inactive, it will not be shown anywhere.
2. Disable: when the button is inactive, it will be shown as disabled.

Exception: in the dropdown, all buttons will always be shown (just potentially disabled).

# Testing

Added an example that shows all features of the buttons in action. Also, added unit tests that test the features as described above. 
![button-links](https://user-images.githubusercontent.com/7528512/74865229-d2dc5200-531e-11ea-815f-ca851548f5cf.gif)

jakedsouza commented:
# [Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=h1) Report
> Merging [#34](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=desc) into [master](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/commit/1013ce492be45fe51ed3de4d3ecc128858ef7790&el=desc) will **decrease** coverage by `1.39%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34/graphs/tree.svg?width=650&height=150&src=pr&token=5AHsP51mSf)](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master      #34      +/-   ##
==========================================
- Coverage   84.14%   82.75%   -1.40%     
==========================================
  Files          33       34       +1     
  Lines         492      516      +24     
  Branches       86       89       +3     
==========================================
+ Hits          414      427      +13     
- Misses         54       66      +12     
+ Partials       24       23       -1     
```


| [Impacted Files](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [...tagrid/renderers/linked-text-renderer.component.ts](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34/diff?src=pr&el=tree#diff-cHJvamVjdHMvY29tcG9uZW50cy9zcmMvZGF0YWdyaWQvcmVuZGVyZXJzL2xpbmtlZC10ZXh0LXJlbmRlcmVyLmNvbXBvbmVudC50cw==) | `100.00% <0.00%> (ø)` | |
| [...ects/components/src/datagrid/datagrid.component.ts](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34/diff?src=pr&el=tree#diff-cHJvamVjdHMvY29tcG9uZW50cy9zcmMvZGF0YWdyaWQvZGF0YWdyaWQuY29tcG9uZW50LnRz) | `79.56% <0.00%> (-9.62%)` | :arrow_down: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=footer). Last update [1013ce4...e211738](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

juanmendes commented on projects/components/src/datagrid/datagrid.component.html:
```diff
@@ -1,4 +1,60 @@
+        </div>
+
+        <ng-container
+            *ngIf="
```
Avoid logic  in templates. See https://angular.io/guide/template-syntax#expression-guidelines

> A property name or method call should be the norm, but an occasional Boolean negation, !, is OK

Please fix throughout change

juanmendes commented on projects/components/src/datagrid/datagrid.component.ts:
```diff
@@ -274,6 +261,41 @@ export class DatagridComponent<R> implements OnInit {
+                const button = buttons[0];
+                if (button.shouldDisplay(selection)) {
+                    toOutput.push(button);
+                    toTake -= 1;
```
Instead of modifying `toTake`, can't we just always compare the length of `toOutput`?

juanmendes commented on projects/components/src/datagrid/datagrid.component.spec.ts:
```diff
@@ -350,6 +356,134 @@ describe('DatagridComponent', () => {
+                });
+
+                it('shows the correct buttons', function(this: HasFinderAndGrid): void {
+                    expect(this.clrGridWidget.getButtons()).toEqual([' Add ']);
```
space around the text is fishy

juanmendes replied:
What does correct buttons mean? Shows buttons where shouldDisplay returns true?

juanmendes commented on projects/components/src/datagrid/datagrid.component.spec.ts:
```diff
@@ -350,6 +356,134 @@ describe('DatagridComponent', () => {
+                        expect(spy).toHaveBeenCalledWith([mockData[0]]);
+                    });
+
+                    it('calculates the correct number of buttons to show on a row', function(this: HasFinderAndGrid): void {
```
Please avoid correct. What does it mean to be correct? The text of the it should try to match the text of the expect.  

it adds CSS classes to set the correct width based on the number of buttons

juanmendes commented on projects/components/src/datagrid/datagrid.component.ts:
```diff
@@ -274,6 +261,41 @@ export class DatagridComponent<R> implements OnInit {
+                    toTake -= 1;
+                }
+            } else {
+                throw new Error('Featured button was not found');
```
I don't think this line is covered in tests. I also didn't see a test that made sure a button that was supposed to be featured but is not available

juanmendes commented on projects/components/src/datagrid/interfaces/datagrid-column.interface.ts:
```diff
@@ -40,6 +40,106 @@ export enum GridColumnSortDirection {
+     */
+    class?: string;
+    /**
+     * The unique ID of this button.
```
Only unique within the buttons the user has passed in

juanmendes commented on projects/components/src/datagrid/interfaces/datagrid-column.interface.ts:
```diff
@@ -40,6 +40,106 @@ export enum GridColumnSortDirection {
+     */
+    handler: (entity: R[]) => void;
+    /**
+     * The function that is called to determine if the button should be displayed.
```
This may be shown as disabled instead of hidden. The intended behavior is disabled, we may need a switch if we want this to be configurable. Let's chat with Martin

ryan-bradford replied:
Updated as per the comment with Martin in the most recent commit. Tell me what you think!

juanmendes commented on projects/components/src/datagrid/interfaces/datagrid-column.interface.ts:
```diff
@@ -40,6 +40,106 @@ export enum GridColumnSortDirection {
+ */
+export interface ButtonConfig<R> {
+    /**
+     * The buttons who's displayability does not depend on the selected entity.
```
whose here and below

juanmendes commented on projects/components/src/utils/test/datagrid/datagrid.wo.ts:
```diff
@@ -124,6 +124,54 @@ export class ClrDatagridWidgetObject extends WidgetObject<ClrDatagrid> {
+     * Presses the contextual button at the given {@param index}.
+     */
+    pressContextualButton(index: number): void {
+        this.click(`clr-dg-action-bar:nth-of-type(1n) button:nth-of-type(${index + 1}n)`);
```
Shouldn't need to specify first. It's the default. What does 1n mean?

ps37 commented on projects/components/src/datagrid/interfaces/datagrid-column.interface.ts:
```diff
@@ -40,6 +40,106 @@ export enum GridColumnSortDirection {
+/**
+ * A type of button who's displayability dependends on the selected entity.
+ */
+export interface ContextualButton<R> {
```
GlobalButton and ContextualButton seems to have same properties except for id and Icon. How about making a common interface like...
GridButton {
 ...common properties
 shouldDisplay: (recs?: R[]) => boolean;
 handler(entity?: R[]) => void;
}

ContextualButton can then extend the interface

ryan-bradford replied:
I extracted the label and class to the common interface. I chose not to extract the handler or shouldDisplay function to the interface because that would require giving GlobalButton the ability to accept records in the callback functions and a generic type which felt a little awkward. What do you think about this?

ps37 commented on projects/components/src/datagrid/interfaces/datagrid-column.interface.ts:
```diff
@@ -40,6 +40,106 @@ export enum GridColumnSortDirection {
+     *
+     * @param entity the currently selected entities.
+     */
+    shouldDisplay: (rec: R[]) => boolean;
```
nit; I feel this would be better if it's named as isVisible or isDisabled

ryan-bradford replied:
Related to Juan's above comment about talking with Martin. Will update this when that conversation is resolved

ryan-bradford replied:
changes to isActive in the most recent commit

ps37 commented on projects/components/src/datagrid/datagrid.component.html:
```diff
@@ -1,4 +1,60 @@
+        >
+            <div class="btn-group" *ngIf="this.getFeaturedButtons(this.getDatagridSelection()).length !== 0">
+                <ng-container *ngFor="let button of this.getFeaturedButtons(this.getDatagridSelection())">
+                    <button type="button" class="btn btn-icon" (click)="button.handler(getDatagridSelection())">
```
The signature of `getDatagridSelection` method in the TS below can be changed to `get datagridSelection`. This way calls to the method can be a little less verbose as `button.handler(datagridSelection)`

ps37 commented on projects/components/src/datagrid/datagrid.component.html:
```diff
@@ -1,4 +1,60 @@
+                getDatagridSelection().length !== 0
+            "
+        >
+            <div class="btn-group" *ngIf="this.getFeaturedButtons(this.getDatagridSelection()).length !== 0">
```
Can this if condition be merged with the above if condition and can we get rid off the <ng-container> at line 29?

ryan-bradford replied:
This condition needs to be separate from the above condition because we always want to show the `dropdown`, but only want to show the `btn-group` if there are buttons to put in it.

ps37 commented on projects/components/src/datagrid/interfaces/datagrid-column.interface.ts:
```diff
@@ -40,6 +40,106 @@ export enum GridColumnSortDirection {
+     */
+    featured: string[];
+    /**
+     * How many buttons should display on the featured section.
```
How is that different from the length of featured list above? Can you please add that to the comment?

ryan-bradford replied:
Tried to clarify this, let me know what you think

ps37 replied:
Sorry, I don't understand the comment. If I remember right... you said this property is to select first 'n' non-disabled buttons from the featured list. And you said that's necessary to save the screen's real state.

ps37 replied:
I am sorry I don't understand the comment. As I remember, you said that it is for saving horizontal real estate on the screen by choosing that many number of non-disabled featured buttons from the featured list. Right?

ps37 commented on projects/examples/src/components/datagrid/datagrid-link.example.component.ts:
```diff
@@ -0,0 +1,131 @@
+        },
+    };
+
+    // Always disable for both options
```
Should these comments be removed?

ryan-bradford commented on package.json:
```diff
@@ -64,6 +64,7 @@
     "@angular/cli": "8.3.9",
     "@angular/compiler-cli": "8.2.8",
     "@angular/language-service": "8.2.8",
+    "@compodoc/compodoc": "^1.1.11",
```
Fixed in a later commit

ryan-bradford commented on projects/components/src/datagrid/datagrid.component.ts:
```diff
@@ -274,6 +261,41 @@ export class DatagridComponent<R> implements OnInit {
+     * Returns the buttons that should be featured given the {@param selection} of currently selected entities.
+     */
+    getFeaturedButtons(selection: R[]): ContextualButton<R>[] {
+        let toTake = this.buttonConfig.contextualButtons.featuredCount;
```
This would work, except for the logic that if a featured button is disabled, we want to skip it and find another. Because of this, we might need to check more `featured` buttons than just `.slice(0, featuredButtonCount)` would give.

ryan-bradford commented:
Responded to both reviews. Something got screwed up with the pull request, so I had to force push a change, but the most recent commit still reflects all the relevant review changes.

ps37 commented on projects/components/src/datagrid/interfaces/datagrid-column.interface.ts:
```diff
@@ -114,10 +112,12 @@ export interface ContextualButtonConfig<R> {
     featured: string[];
     /**
      * How many buttons should display on the featured section.
+     * Allows the user to specific more {@link featured} buttons than necessary and for some to act as fallbacks.
```
Sorry, I don't understand the comment. If I remember right... you said this property is to select first 'n' non-disabled buttons from the featured list. And you said that's necessary to save the screen's real state.

juanmendes commented on projects/components/src/datagrid/datagrid.component.html:
```diff
@@ -1,4 +1,42 @@
+                *ngIf="isButtonShown(button)"
+                (click)="button.handler()"
+                [disabled]="isButtonDisabled(button, button.isActive())"
+            >
```
This should not have spaces around the text, then you can fix any `trim()`s in tests
```typescript
>{{
 button.label 
}}</button>
```

juanmendes commented on projects/components/src/utils/test/datagrid/datagrid.wo.ts:
```diff
@@ -124,6 +124,54 @@ export class ClrDatagridWidgetObject extends WidgetObject<ClrDatagrid> {
+     * Gives a list of the labels of the displayed action buttons next to the given row.
+     */
+    getRowButtons(rowIndex: number): string[] {
+        return this.findElements(`clr-dg-row:nth-of-type(${rowIndex + 1}) .action-button-group button`).map(
```
We should be able to pass `this.rows[i]`as the parent context for the search

juanmendes commented on projects/examples/src/components/datagrid/datagrid-link.example.component.ts:
```diff
@@ -0,0 +1,135 @@
+    GridSelectionType,
+    ContextualButtonPosition,
+} from '@vcd/ui-components';
+import { InactiveButtonDisplayMode } from 'projects/components/src/datagrid';
```
We should never directly import from a different project, always through  '@vcd/ui-components'. It often doesn't matter, but it creates two different classes and I've had problems before. We should add a tslint rule to prevent this...

juanmendes commented on projects/components/src/datagrid/datagrid.component.ts:
```diff
@@ -272,7 +272,87 @@ export class DatagridComponent<R> implements OnInit {
      * If the record has a href, defaults to that. Else, defaults to index.
      */
     @Input() trackBy: TrackByFunction<R> = (index: number, record: (R & HasHref) | undefined): string | number =>
```
This is too complex to be inline

Suggestion
```
@Input() trackBy: TrackByFunction<R extends {href?: string}> = (index: number, record: R): string  => {
  return record.href || index.toString();
}
```

ryan-bradford commented:
Updated per Juan's comments!

ryan-bradford commented on projects/components/src/datagrid/datagrid.component.scss:
```diff
@@ -3,16 +3,41 @@
+// The amount of padding that should be placed above and below a button group.
+$PADDING: 5px;
+// How wide a single button should be.
+$BUTTON_WIDTH: 50px;
```
Didn't resolve this exactly through changing the width of the button, but by correcting the padding + width on the button group CSS class

ryan-bradford commented:
Updated as per review. I also changed buttonConfig.inactiveButtonDisplayMode to an optional property that defaults to Disabled. I did this because it makes the tests look cleaner and the user experience simpler.

juanmendes commented on projects/components/src/datagrid/datagrid.component.html:
```diff
@@ -1,6 +1,6 @@
 <clr-datagrid [clrDgLoading]="isLoading" [ngClass]="this.clrDatagridCssClass" (clrDgRefresh)="gridStateChanged($event)">
-    <clr-dg-action-bar *ngIf="buttonConfig">
-        <div class="btn-group" *ngFor="let button of buttonConfig.globalButtons">
+    <clr-dg-action-bar *ngIf="_buttonConfig">
```
Please make this private and create a getter for it

jakedsouza commented:
# [Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=h1) Report
> Merging [#34](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=desc) into [master](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/commit/21638d1ca6455957bd525b2227f88885e0ff9a23&el=desc) will **increase** coverage by `1.51%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34/graphs/tree.svg?width=650&height=150&src=pr&token=5AHsP51mSf)](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master      #34      +/-   ##
==========================================
+ Coverage   84.38%   85.89%   +1.51%     
==========================================
  Files          33       34       +1     
  Lines         506      546      +40     
  Branches       88       95       +7     
==========================================
+ Hits          427      469      +42     
+ Misses         55       54       -1     
+ Partials       24       23       -1     
```


| [Impacted Files](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [...tagrid/renderers/linked-text-renderer.component.ts](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34/diff?src=pr&el=tree#diff-cHJvamVjdHMvY29tcG9uZW50cy9zcmMvZGF0YWdyaWQvcmVuZGVyZXJzL2xpbmtlZC10ZXh0LXJlbmRlcmVyLmNvbXBvbmVudC50cw==) | `100.00% <0.00%> (ø)` | |
| [...ects/components/src/datagrid/datagrid.component.ts](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34/diff?src=pr&el=tree#diff-cHJvamVjdHMvY29tcG9uZW50cy9zcmMvZGF0YWdyaWQvZGF0YWdyaWQuY29tcG9uZW50LnRz) | `93.80% <0.00%> (+4.39%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=footer). Last update [21638d1...47714da](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/34?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).
